### PR TITLE
Test cases for each API

### PR DESCRIPTION
**Pull request by ShrijanL**
- Created test separate file for each API with parameterization.

### CODE REVIEW

File: django_generic_api/tests/test_fetch.py [hunk 1]
```diff
@@ -0,0 +1,1213 @@
...
+            headers=headers,
+        )
+        response_data = json.loads(response.content.decode("utf-8"))
+        assert response.status_code == 200
+        assert response_data["total"] == 1
+        assert response_data["data"] == [
+            {"name": "test_user1", "email": "user1@gmail.com"}
+        ]
+
+    def test_wrong_payload_format(
```
Comment (nareshelequin) [inline django_generic_api/tests/test_fetch.py:162]: test_invalid_payload_format

File: django_generic_api/tests/test_fetch.py [hunk 2]
```diff
@@ -0,0 +1,1213 @@
...
+        )
+        response_data = json.loads(response.content.decode("utf-8"))
+        assert response.status_code == 400
+        assert response_data["code"] == "DGA-V006"
+        assert (
+            response_data["error"]
+            == "Input should be a valid string('filters', 0, 'name')"
+        )
+
+    def test_invalid_fetch_filter_name_field(
```
Comment (nareshelequin) [inline django_generic_api/tests/test_fetch.py:520]: repeated test case.

File: django_generic_api/tests/test_fetch.py [hunk 3]
```diff
@@ -0,0 +1,1213 @@
...
+        )
+        response_data = json.loads(response.content.decode("utf-8"))
+        assert response.status_code == 400
+        assert response_data["code"] == "DGA-V009"
+        assert (
+            response_data["error"]
+            == "{'error': \"Extra field {'ABCD'}\", 'code': 'DGA-U002'}"
+        )
+
+    def test_invalid_fetch_filter_value_datatype(
```
Comment (nareshelequin) [inline django_generic_api/tests/test_fetch.py:561]: test_invalid_fetch_filter_value_element_datatype

File: django_generic_api/tests/test_fetch.py [hunk 4]
```diff
@@ -0,0 +1,1213 @@
...
+        )
+        response_data = json.loads(response.content.decode("utf-8"))
+        assert response.status_code == 400
+        assert response_data["code"] == "DGA-V009"
+        assert (
+            response_data["error"]
+            == "{'error': \"Invalid data: ['456789'] for dob\", 'code': 'DGA-S004'}"
+        )
+
+    def test_multiple_filter_eq_operator(
```
Comment (nareshelequin) [inline django_generic_api/tests/test_fetch.py:598]: test_invalid_filter_value_length_for_eq_operator

File: django_generic_api/tests/test_fetch.py [hunk 5]
```diff
@@ -0,0 +1,1213 @@
...
+        response_data = json.loads(response.content.decode("utf-8"))
+
+        assert response.status_code == 400
+        assert response_data["code"] == "DGA-V006"
+        assert (
+            response_data["error"]
+            == "Input should be 'or' or 'and'('filters', 0, 'operation')"
+        )
+
+    def test_pagesize_is_not_int(
```
Comment (nareshelequin) [inline django_generic_api/tests/test_fetch.py:683]: test_unknown_page_number

File: django_generic_api/tests/test_fetch.py [hunk 6]
```diff
@@ -0,0 +1,1213 @@
...
+            headers=headers,
+        )
+
+        response_data = json.loads(response.content.decode("utf-8"))
+
+        assert response.status_code == 400
+        assert response_data["code"] == "DGA-V009"
+        assert response_data["error"] == "Negative indexing is not supported."
+
+    def test_sort_not_dictionary(
```
Comment (nareshelequin) [inline django_generic_api/tests/test_fetch.py:764]: test_unknow_format_sort_property

File: django_generic_api/tests/test_fetch.py [hunk 7]
```diff
@@ -0,0 +1,1213 @@
...
+        )
+        response_data = json.loads(response.content.decode("utf-8"))
+        assert response.status_code == 400
+        assert response_data["code"] == "DGA-V006"
+        assert (
+            response_data["error"]
+            == "Input should be a valid boolean, unable to interpret input('distinct',)"
+        )
+
+    def test_missing_authentication_header(self, fetch_data_1, api_client):
```
Comment (nareshelequin) [inline django_generic_api/tests/test_fetch.py:945]: repeated test case.

File: django_generic_api/tests/test_fetch.py [hunk 8]
```diff
@@ -0,0 +1,1213 @@
...
+            "/api/fetch/",
+            fetch_payload,
+            format="json",
+        )
+        response_data = json.loads(response.content.decode("utf-8"))
+        assert response.status_code == 401
+        assert response_data["code"] == "DGA-S001"
+        assert response_data["error"] == "Unauthorized access"
+
+    def test_incorrect_token_format(
```
Comment (nareshelequin) [inline django_generic_api/tests/test_fetch.py:974]: repeated test case.

File: django_generic_api/tests/test_fetch.py [hunk 9]
```diff
@@ -0,0 +1,1213 @@
...
+            format="json",
+            headers=headers,
+        )
+        response_data = json.loads(response.content.decode("utf-8"))
+        assert response.status_code == 401
+        assert response_data["code"] == "DGA-S002"
+        assert response_data["error"] == "Invalid Token"
+
+    def test_missing_view_permissions(
+        self, fetch_data_1, api_client, add_perm_token
```
Comment (nareshelequin) [inline django_generic_api/tests/test_fetch.py:1010]: repeated test case.

File: django_generic_api/tests/test_fetch.py [hunk 10]
```diff
@@ -0,0 +1,1213 @@
...
+        )
+        response_data = json.loads(response.content.decode("utf-8"))
+        assert response.status_code == 404
+        assert response_data["code"] == "DGA-V008"
+        assert (
+            response_data["error"]
+            == "Something went wrong!!! Please contact the administrator."
+        )
+
+    def test_positive_fetch_scenario(
```
Comment (nareshelequin) [inline django_generic_api/tests/test_fetch.py:1047]: test_fetch_single_filter

File: django_generic_api/tests/test_login.py [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+        # Sending POST request to login endpoint
+        response = api_client.post("/api/login/", login_payload, format="json")
+        response_data = json.loads(response.content.decode("utf-8"))
+
+        # Assertions
+        assert response.status_code == 200
+        assert "refresh" in response_data["data"][0]
+        assert "access" in response_data["data"][0]
+
+    def test_missing_field_in_payload(self, api_client):
```
Comment (nareshelequin) [inline django_generic_api/tests/test_login.py:34]: test_missing_username_property

**Review by nareshelequin — CHANGES_REQUESTED** [review]

File: django_generic_api/tests/test_activate.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+class TestAccountActivateAPI:
+
+    def test_expired_activation_link(
+        self, api_client, email_activate_inactive_user_id
+    ):
+        """
+        User's activation link is expired.
+        """
+        timestamp = int(time.time()) - (25 * 60 * 60)  # 25 hours old timestamp
+        token = f"{email_activate_inactive_user_id}:{timestamp}"
```
[the remark this answers is not in the document]
Comment (ShrijanL) [reply django_generic_api/tests/test_activate.py:24]: encoded the token using base64 to not reveal the information in token.

File: django_generic_api/tests/test_activate.py
[the remark this answers is not in the document]
Comment (ShrijanL) [reply django_generic_api/tests/test_activate.py:1]: Changed file name to test_activate_email.py

File: django_generic_api/tests/test_activate.py [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+    expected_status,
+    expected_message,
+    expected_error,
+    expected_code,
+):
+    # Retrieve or prepare user ID based on the setup fixture
+    user_id = request.getfixturevalue(setup_user)
+
+    if expected_error == "User not found.":
+        User.objects.filter(id=user_id).delete()  # Simulate user deletion
```
[the remark this answers is not in the document]
Comment (ShrijanL) [reply django_generic_api/tests/test_activate.py:139]: removed as passed a non existing user.

File: django_generic_api/tests/test_fetch.py [hunk 14]
```diff
@@ -0,0 +1,1213 @@
...
+    save_perm_user,
+)
+from unittest import mock
+from rest_framework_simplejwt.tokens import AccessToken
+
+
+@pytest.mark.django_db
+class TestGenericFetchAPI:
+
+    def test_missing_required_field(
```
[the remark this answers is not in the document]
Comment (ShrijanL) [reply django_generic_api/tests/test_fetch.py:21]: renamed.

File: django_generic_api/tests/test_fetch.py [hunk 15]
```diff
@@ -0,0 +1,1213 @@
...
+        }
+
+        # Send a POST request to the fetch endpoint without authentication
+        response = api_client.post("/api/fetch/", fetch_payload, format="json")
+        response_data = json.loads(response.content.decode("utf-8"))
+        assert response.status_code == 401
+        assert response_data["error"] == "Unauthorized access"
+        assert response_data["code"] == "DGA-S001"
+
+    def test_wrong_token_format(self, api_client, view_perm_token):
```
[the remark this answers is not in the document]
Comment (ShrijanL) [reply django_generic_api/tests/test_fetch.py:88]: renamed.

File: django_generic_api/tests/test_fetch.py [hunk 16]
```diff
@@ -0,0 +1,1213 @@
...
+            fetch_payload,
+            format="json",
+            headers=headers,
+        )
+        response_data = json.loads(response.content.decode("utf-8"))
+        assert response.status_code == 401
+        assert response_data["error"] == "Invalid Token"
+        assert response_data["code"] == "DGA-S002"
+
+    def test_fetch_perm_user(self, fetch_data_1, api_client, view_perm_token):
```
[the remark this answers is not in the document]
Comment (ShrijanL) [reply django_generic_api/tests/test_fetch.py:124]: renamed.

File: django_generic_api/tests/test_fetch.py (hunk 1 above)
Comment (ShrijanL) [reply django_generic_api/tests/test_fetch.py:162]: renamed.

File: django_generic_api/tests/test_fetch.py [hunk 17]
```diff
@@ -0,0 +1,1213 @@
...
+        )
+        response_data = json.loads(response.content.decode("utf-8"))
+        assert response.status_code == 404
+        assert (
+            response_data["error"]
+            == "Something went wrong!!! Please contact the administrator."
+        )
+        assert response_data["code"] == "DGA-V008"
+
+    def test_modelName_not_str(
```
[the remark this answers is not in the document]
Comment (ShrijanL) [reply django_generic_api/tests/test_fetch.py:280]: renamed,

File: django_generic_api/tests/test_fetch.py [hunk 18]
```diff
@@ -0,0 +1,1213 @@
...
+        )
+        response_data = json.loads(response.content.decode("utf-8"))
+        assert response.status_code == 400
+        assert (
+            response_data["error"]
+            == "Input should be a valid string('modelName',)"
+        )
+        assert response_data["code"] == "DGA-V006"
+
+    def test_wrong_fields_datatype(
```
[the remark this answers is not in the document]
Comment (ShrijanL) [reply django_generic_api/tests/test_fetch.py:321]: renamed,

File: django_generic_api/tests/test_fetch.py (hunk 2 above)
Comment (ShrijanL) [reply django_generic_api/tests/test_fetch.py:520]: removed test case.

File: django_generic_api/tests/test_fetch.py (hunk 3 above)
Comment (ShrijanL) [reply django_generic_api/tests/test_fetch.py:561]: renamed.

File: django_generic_api/tests/test_fetch.py (hunk 5 above)
Comment (ShrijanL) [reply django_generic_api/tests/test_fetch.py:683]: renamed.

File: django_generic_api/tests/test_fetch.py (hunk 6 above)
Comment (ShrijanL) [reply django_generic_api/tests/test_fetch.py:764]: renamed.

File: django_generic_api/tests/test_fetch.py (hunk 7 above)
Comment (ShrijanL) [reply django_generic_api/tests/test_fetch.py:945]: removed test case.

File: django_generic_api/tests/test_fetch.py (hunk 9 above)
Comment (ShrijanL) [reply django_generic_api/tests/test_fetch.py:1010]: removed test case.

File: django_generic_api/tests/test_fetch.py (hunk 8 above)
Comment (ShrijanL) [reply django_generic_api/tests/test_fetch.py:974]: removed test case,

File: django_generic_api/tests/test_fetch.py (hunk 10 above)
Comment (ShrijanL) [reply django_generic_api/tests/test_fetch.py:1047]: renamed.

File: django_generic_api/tests/test_login.py (hunk 11 above)
Comment (ShrijanL) [reply django_generic_api/tests/test_login.py:34]: renamed.

File: django_generic_api/apps.py [hunk 19]
```diff
@@ -19,7 +19,7 @@ def ready(self):
             ],
             "DEFAULT_THROTTLE_RATES": {
                 "user": "2000/hour",  # Rate limit for authenticated users
-                "anon": "20/hour",  # Rate limit for unauthenticated users,
+                "anon": "25/hour",  # Rate limit for unauthenticated users,
```
Comment (nareshelequin) [inline django_generic_api/apps.py:22]: should be configurable, as per USER request.

**Review by nareshelequin — CHANGES_REQUESTED** [review]

File: django_generic_api/apps.py (hunk 19 above)
Comment (ShrijanL) [reply django_generic_api/apps.py:22]: created ticket for this issue.
https://github.com/jknaresh/django_generic_api/issues/30

**Review by nareshelequin — APPROVED** [review]

**Pull request merged**